### PR TITLE
[CARBONDATA-241]Fixed out of memory issue during query execution

### PR DESCRIPTION
**Pull request by kumarvishal09**
**Problem:** During long run query execution is taking more time and it is throwing out of memory issue.
**Reason**: In compaction we are compacting segments and each segment metadata is loaded in memory. So after compaction compacted segments are invalid but its meta data is not removed from memory because of this duplicate metadata is pile up and it is taking more memory and after few days query execution is throwing OOM
**Solution**: Need to remove invalid blocks from memory

### CODE REVIEW

**Comment by kumarvishal09** [comment]
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/209/

**Comment by kumarvishal09** [comment]
handled build failure issue as a part of this PR

File: processing/src/main/java/org/apache/carbondata/lcm/status/SegmentStatusManager.java [hunk 1]
```diff
@@ -102,6 +91,60 @@ public long getTableStatusLastModifiedTime() throws IOException {
 
   /**
    * get valid segment for given table
+   *
+   * @return
+   * @throws IOException
+   */
+  public InvalidSegmentsInfo getInvalidSegments() throws IOException {
```
Comment (gvramana) [inline processing/src/main/java/org/apache/carbondata/lcm/status/SegmentStatusManager.java]: This requires reading SegmentInfo twice, once for valid blocks and next for Invalid Blocks. Instead send a single class containting ValidAndInvalidBlocks.

File: integration/spark/src/main/scala/org/apache/spark/sql/CarbonDatasourceHadoopRelation.scala [hunk 2]
```diff
@@ -20,20 +20,13 @@ package org.apache.spark.sql
 import java.text.SimpleDateFormat
 import java.util.Date
 
-import org.apache.carbondata.core.carbon.AbsoluteTableIdentifier
```
Comment (gvramana) [inline integration/spark/src/main/scala/org/apache/spark/sql/CarbonDatasourceHadoopRelation.scala]: Merged this commit (compilation issue) changes separately. So can take out those changes from PR.

File: processing/src/main/java/org/apache/carbondata/lcm/status/SegmentStatusManager.java (hunk 1 above)
Comment (kumarvishal09) [inline processing/src/main/java/org/apache/carbondata/lcm/status/SegmentStatusManager.java]: ok i will handle

**Comment by kumarvishal09** [comment]
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/224/

File: integration/spark/src/main/scala/org/apache/spark/sql/CarbonDatasourceHadoopRelation.scala (hunk 2 above)
Comment (kumarvishal09) [inline integration/spark/src/main/scala/org/apache/spark/sql/CarbonDatasourceHadoopRelation.scala]: done

**Comment by gvramana** [comment]
Invalid segment deletion, need not be through CarbonInputFormat, When Invalid segments list given to Btree(both in Driver and executor it should able it delete invalid blocks).

File: hadoop/src/main/java/org/apache/carbondata/hadoop/CarbonInputFormat.java [hunk 3]
```diff
@@ -101,6 +106,8 @@
   //comma separated list of input segment numbers
   public static final String INPUT_SEGMENT_NUMBERS =
       "mapreduce.input.carboninputformat.segmentnumbers";
+  public static final String INVALID_SEGMENT_NUMBERS =
+      "mapreduce.input.carboninputformat.invalidsegmentnumbers";
```
Comment (gvramana) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/CarbonInputFormat.java]: Invalid segment deletion, need not be through CarbonInputFormat, When Invalid segments list given to Btree(both in Driver and executor it should able it delete invalid blocks).

File: core/src/main/java/org/apache/carbondata/core/carbon/datastore/BlockIndexStore.java [hunk 4]
```diff
@@ -260,11 +295,29 @@ public void removeTableBlocks(List<TableBlockInfo> removeTableBlocksInfos,
...
+      for (String segmentId : segmentsToBeRemoved) {
+        List<TableBlockInfo> tableBlockInfoList = segmentIdToBlockInfoMap.get(segmentId);
+        if (null == tableBlockInfoList) {
+          continue;
+        }
+        Iterator<TableBlockInfo> tableBlockInfoIterator = tableBlockInfoList.iterator();
+        while (tableBlockInfoIterator.hasNext()) {
+          TableBlockInfo info = tableBlockInfoIterator.next();
+          AbstractIndex remove = map.remove(info);
+          if (null != remove) {
```
Comment (gvramana) [inline core/src/main/java/org/apache/carbondata/core/carbon/datastore/BlockIndexStore.java]: tableBlockInfoIterator.remove needs to called irrespective of null != remove

**Comment by kumarvishal09** [comment]
@gvramana fixed all the review comments, Please check

**Comment by kumarvishal09** [comment]
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/243/

File: hadoop/src/test/java/org/apache/carbondata/hadoop/ft/CarbonInputMapperTest.java [hunk 5]
```diff
@@ -129,6 +132,37 @@ private int countTheColumns(String outPath) throws Exception {
     return 0;
   }
 
+  private void runJob(String outPath, CarbonProjection projection, Expression filter)
```
Comment (gvramana) [inline hadoop/src/test/java/org/apache/carbondata/hadoop/ft/CarbonInputMapperTest.java]: move back the code to original place

File: hadoop/src/main/java/org/apache/carbondata/hadoop/CarbonInputFormat.java [hunk 6]
```diff
@@ -706,8 +725,9 @@ private String getUpdateExtension() {
   /**
    * @return updateExtension
    */
-  private String[] getValidSegments(JobContext job) throws IOException {
-    String segmentString = job.getConfiguration().get(INPUT_SEGMENT_NUMBERS, "");
+  private String[] getSegmentsFromConfiguration(JobContext job, String segmentType)
+      throws IOException {
+    String segmentString = job.getConfiguration().get(segmentType, "");
```
Comment (gvramana) [inline hadoop/src/main/java/org/apache/carbondata/hadoop/CarbonInputFormat.java]: change signature to previous one

File: integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonScanRDD.scala [hunk 7]
```diff
@@ -102,7 +115,7 @@ class CarbonScanRDD[V: ClassTag](
     val splits = carbonInputFormat.getSplits(job)
     if (!splits.isEmpty) {
       val carbonInputSplits = splits.asScala.map(_.asInstanceOf[CarbonInputSplit])
-
+      queryModel.setInvalidSegmentIds(validAndInvalidSegments.getInvalidSegments)
```
Comment (gvramana) [inline integration/spark/src/main/scala/org/apache/carbondata/spark/rdd/CarbonScanRDD.scala]: move this to common getSplits, other wise validAndInvalidSegments can be null, if parallel deletion happens.

**Comment by gvramana** [comment]
http://136.243.101.176:8080/job/ApacheCarbonManualPRBuilder/252/
LGTM

**Pull request closed without merging**